pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java

```diff
@@ -349,7 +350,11 @@ protected void readMoreEntries(Consumer consumer) {
                 }
                 havePendingRead = true;
                 if (consumer.readCompacted()) {
-                    boolean readFromEarliest = isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId());
+                    boolean readFromEarliest = false;
+                    if (!cursor.isDurable() || ((ManagedCursorImpl) cursor).isCompactionCursor()
+                            || cursor.getPersistentMarkDeletedPosition() == null) {
+                        readFromEarliest =  isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId());
```

Comment:
The expected behaviors from the users point:
- Create a new reader. It should be reading at the earliest entry.
  - Should continue reading from `{markDeletedPosition} + 1` after reconnect(caused by socket closed or other cases)<sup>[1]</sup>
- Create a new consumer(`non-durable`). It should be reading at the earliest entry.
  - Should continue reading from `{markDeletedPosition} + 1` after reconnect(caused by socket closed or other cases)<sup>[2]</sup>
- Create a new consumer(`durable`). It should be reading at `{markDeletedPosition} + 1.`<sup>[3]</sup>

This PR only fixes the case<sup>[3]</sup>, we should guarantee the case<sup>[1]</sup> and case<sup>[2]</sup> too.

Reply:
For `non-durable` consumer, it seems the consumer should read from startMessageId after reconnecting.

Refer to: https://github.com/apache/pulsar/blob/7a3f304d6b6f658ef825b6464f192b42e4833a26/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L803-L812

Reply:
Got it.  I wrote a test below(it works as expected). Could you add it into the current PR to guarantee this feature will not be broken?

```java
@Test
public void testReaderReconnectedFromNextEntry() throws Exception {
    final String topic = newTopicName();
    Reader<String> reader = pulsarClient.newReader(Schema.STRING).topic(topic).receiverQueueSize(1)
            .startMessageId(MessageId.earliest).create();
    Producer<String> producer = pulsarClient.newProducer(Schema.STRING).topic(topic).create();

    // Send 3 and consume 1.
    producer.send("1");
    producer.send("2");
    producer.send("3");
    Message<String> msg1 = reader.readNext(2, TimeUnit.SECONDS);
    assertEquals(msg1.getValue(), "1");

    // Trigger reader reconnect.
    admin.topics().unload(topic);

    // For non-durable we are going to restart from the next entry.
    Message<String> msg2 = reader.readNext(2, TimeUnit.SECONDS);
    assertEquals(msg2.getValue(), "2");
    Message<String> msg3 = reader.readNext(2, TimeUnit.SECONDS);
    assertEquals(msg3.getValue(), "3");

    // cleanup.
    reader.close();
    producer.close();
    admin.topics().delete(topic, false);
}
```

Reply:
Thanks, I already add this test into PR.